File: Assemblies/EonicWebV5/Tools/PerfLog.vb
Option Strict Off
Option Explicit On
Imports System.Xml
Imports System.Web.HttpUtility
Imports System.Web.Configuration
Imports System.Web.Mail
Imports System.Reflection
Imports VB = Microsoft.VisualBasic
Imports System.Security.Cryptography
Imports System.Text
Imports System.Net.Mail
Imports System.Diagnostics
Imports System

Public Class PerfLog

    Public cSiteName As String
    Public cDataConn As String = ""

    Private bLoggingOn As Boolean
    Private nStep As Integer
    Private oBuilder As StringBuilder
    Private oPerfMonRequests As PerformanceCounter
    Private Entries(0) As String
    Dim dLast As Date = Now
    Dim nTimeAccumalative As Double = 0
    Dim nMemLast As Integer = 0
    Dim nProcLast As Integer = 0


    ' Counters
    Private _workingSetPrivateMemoryCounter As PerformanceCounter
    Private _workingSetMemoryCounter As PerformanceCounter


    Private moCtx As System.Web.HttpContext = System.Web.HttpContext.Current

    'Session / Request Level Properties
    Public moRequest As System.Web.HttpRequest = moCtx.Request
    Public moResponse As System.Web.HttpResponse = moCtx.Response
    Public moSession As System.Web.SessionState.HttpSessionState = moCtx.Session
    Public moServer As System.Web.HttpServerUtility = moCtx.Server

    Public moConfig As System.Collections.Specialized.NameValueCollection = WebConfigurationManager.GetWebApplicationSection("eonic/web")




    Public Property Logging() As Boolean
        Get
            Return bLoggingOn
        End Get
        Set(ByVal value As Boolean)
            bLoggingOn = value
        End Set
    End Property

    Public Sub New(ByVal SiteName As String)
        Try
            cSiteName = SiteName
            If Not moCtx.Session Is Nothing Then moSession = moCtx.Session
            If Not moSession Is Nothing Then
                If moSession("Logging") = "On" Then
                    TurnOn()
                End If
            End If
        Catch ex As Exception

        End Try


    End Sub

    Private Sub TurnOn()
        'If bLoggingOn Then Exit Sub
        Try
            If Not bLoggingOn Then
                bLoggingOn = True
                nStep = 0
                moSession("Logging") = "On"
                Try
                    ' oPerfMonRequests = New System.Diagnostics.PerformanceCounter("ASP.NET v4.0.30319", "Requests Current")
                Catch ex As Exception
                    'do nothing
                End Try

                Dim cSessionRequest As String = moSession("SessionRequest")
                If IsNumeric(cSessionRequest) Then
                    moSession("SessionRequest") = CInt(cSessionRequest) + 1
                    dLast = Now
                    nTimeAccumalative = 0
                    nMemLast = 0
                    nProcLast = 0
                Else
                    moSession("SessionRequest") = 0
                    dLast = Now
                    nTimeAccumalative = 0
                    nMemLast = 0
                    nProcLast = 0
                End If

                _workingSetPrivateMemoryCounter = New PerformanceCounter("Process", "Working Set - Private", Process.GetCurrentProcess.ProcessName)
                _workingSetMemoryCounter = New PerformanceCounter("Process", "Working Set", Process.GetCurrentProcess.ProcessName)
            End If
        Catch ex As Exception
            Debug.WriteLine(ex.ToString)
        End Try
    End Sub

    Public Sub Start()
        TurnOn()
    End Sub

    Public Sub [Stop]()
        bLoggingOn = False
        moSession("Logging") = "Off"
    End Sub

    Public Sub Log(ByVal cModuleName As String, ByVal cProcessName As String, Optional ByVal cDescription As String = "")
        'If Not bLoggingOn Then Exit Sub
        Try
            If bLoggingOn Then

                Dim oLN As TimeSpan = Now - dLast
                nTimeAccumalative += oLN.TotalMilliseconds
                'Dim nMemDif As Integer = Process.GetCurrentProcess.PrivateMemorySize64 - nMemLast
                Dim memoryPrivate As Long
                If _workingSetPrivateMemoryCounter Is Nothing Then
                    memoryPrivate = 0
                Else
                    memoryPrivate = _workingSetPrivateMemoryCounter.NextValue()
                End If

                Dim nMemDif As Long = memoryPrivate - nMemLast
                Dim nProcDif As Long = Process.GetCurrentProcess.PrivilegedProcessorTime.Milliseconds - nProcLast

                nMemLast = memoryPrivate
                nProcLast = Process.GetCurrentProcess.PrivilegedProcessorTime.Milliseconds


                Dim cEntryFull As String = "INSERT INTO tblPerfMon" & _
                " ( MachineName, Website, SessionID, SessionRequest, Path, [Module], [Procedure],Description, Step, [Time],TimeAccumalative, Requests, PrivateMemorySize64, PrivilegedProcessorTimeMilliseconds)" & _
                " VALUES("
                cEntryFull &= "'"
                cEntryFull &= moServer.MachineName & "','"
                cEntryFull &= cSiteName & "','"
                Try
                    cEntryFull &= CStr(moSession.SessionID & "") & "','"
                    cEntryFull &= CStr(moSession("SessionRequest") & "") & "','"
                Catch ex As Exception
                    cEntryFull &= "','','"
                End Try
                'If moSession.SessionID Is Nothing Then

                'Else
                '    cEntryFull &= CStr(moSession.SessionID & "") & "','"
                '    cEntryFull &= CStr(moSession("SessionRequest") & "") & "','"
                'End If
                Dim cPath As String = ""
                If Not System.Web.HttpContext.Current.Request Is Nothing Then
                    cPath = System.Web.HttpContext.Current.Request("path")
                End If
                cEntryFull &= SqlFmt(cPath) & "','"
                cEntryFull &= SqlFmt(cModuleName) & "','"
                cEntryFull &= Left(SqlFmt(cProcessName), 254) & "','"
                cEntryFull &= Left(SqlFmt(cDescription), 3999) & "',"
                cEntryFull &= nStep & ","
                cEntryFull &= oLN.TotalMilliseconds & ","
                cEntryFull &= nTimeAccumalative & ","
                If oPerfMonRequests Is Nothing Then
                    cEntryFull &= "null,'"
                Else
                    cEntryFull &= oPerfMonRequests.RawValue & ",'"
                End If

                cEntryFull &= nMemLast & "','"
                If _workingSetMemoryCounter Is Nothing Then
                    cEntryFull &= ""
                Else
                    cEntryFull &= CLng(_workingSetMemoryCounter.NextValue())
                End If
                'cEntryFull &= Process.GetCurrentProcess.PrivateMemorySize64 & "','"
                'cEntryFull &= Process.GetCurrentProcess.WorkingSet64
                cEntryFull &= "')"
                nStep += 1

                ReDim Preserve Entries(nStep)
                Entries(nStep - 1) = cEntryFull

                'nMemLast = Process.GetCurrentProcess.PrivateMemorySize64
                'nProcLast = Process.GetCurrentProcess.PrivilegedProcessorTime.Milliseconds
                dLast = Now
            Else

                'oPerfMonRequests = New System.Diagnostics.PerformanceCounter("ASP.NET v2.0.50727", "Requests Current")
                'Dim cEntryFull As String = oPerfMonRequests.RawValue

                'Dim oLN As TimeSpan = Now - dLast
                'nTimeAccumalative += oLN.TotalMilliseconds

                Dim nMemDif As Long = Process.GetCurrentProcess.WorkingSet64
                Dim nMemTotal As Long = Process.GetCurrentProcess.PrivateMemorySize64

                'Dim currentProcess As System.Diagnostics.Process = Process.GetCurrentProcess

                'Dim nProcDif As Integer = Process.GetCurrentProcess.PrivilegedProcessorTime.Milliseconds

            End If
        Catch ex As Exception
            Debug.WriteLine(ex.ToString)
        End Try
    End Sub

    Public Sub Write()
        'If Not bLoggingOn Then Exit Sub
        Dim cProcessInfo As String = Nothing
        Try
            If bLoggingOn Then
                Dim ConStr As String = "" 'moConfig("PerfMonConnection")
                If ConStr = "" Then
                    Dim dbAuth As String
                    If moConfig("DatabasePassword") <> "" Then
                        dbAuth = "user id=" & moConfig("DatabaseUsername") & "; password=" & moConfig("DatabasePassword")
                    Else
                        If moConfig("DatabaseAuth") <> "" Then
                            dbAuth = moConfig("DatabaseAuth")
                        Else
                            dbAuth = "Integrated Security=SSPI;"
                        End If
                    End If
                    ConStr = "Data Source=" & moConfig("DatabaseServer") & "; " & _
                        "Initial Catalog=" & moConfig("DatabaseName") & "; " & dbAuth
                End If
                Dim oCon As New SqlClient.SqlConnection(ConStr)
                Dim oCmd As New SqlClient.SqlCommand
                oCmd.Connection = oCon
                oCon.Open()
                Dim i As Integer
                For i = 0 To UBound(Entries)
                    If Not Entries(i) = "" Then
                        cProcessInfo = Entries(i)
                        oCmd.CommandText = Entries(i)
                        oCmd.ExecuteNonQuery()
                    End If
                Next
                oCon.Close()
                bLoggingOn = False
            End If
        Catch ex As Exception

            Debug.WriteLine(cProcessInfo & " - errormsg - " & ex.ToString)
        End Try
    End Sub

End Class

File: Assemblies/Protean.CMS/tools/PerfLog.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Web.Configuration;
using static Protean.stdTools;

namespace Protean
{

    public class PerfLog : IDisposable
    {

        public string cSiteName;
        public string cDataConn = "";

        private bool bLoggingOn;
        private int nStep;
        private PerformanceCounter oPerfMonRequests;
        private List<string> Entries;
        private DateTime dLast = DateTime.Now;
        private double nTimeAccumalative = 0d;
        private int nMemLast = 0;
        private int nProcLast = 0;
        private string LatestLog = "";

        // Shared performance counters (singleton pattern for efficiency)
        private static readonly Lazy<PerformanceCounter> _sharedWorkingSetPrivateMemoryCounter = 
            new Lazy<PerformanceCounter>(() => 
            {
                try
                {
                    return new PerformanceCounter("Process", "Working Set - Private", Process.GetCurrentProcess().ProcessName);
                }
                catch
                {
                    return null;
                }
            });

        private static readonly Lazy<PerformanceCounter> _sharedWorkingSetMemoryCounter = 
            new Lazy<PerformanceCounter>(() => 
            {
                try
                {
                    return new PerformanceCounter("Process", "Working Set", Process.GetCurrentProcess().ProcessName);
                }
                catch
                {
                    return null;
                }
            });

        // Instance references to shared counters
        private PerformanceCounter _workingSetPrivateMemoryCounter;
        private PerformanceCounter _workingSetMemoryCounter;

        // Cached SQL insert prefix
        private static readonly string InsertPrefix = 
            "INSERT INTO tblPerfMon ( MachineName, Website, SessionID, SessionRequest, Path, [Module], [Procedure], Description, Step, [Time], TimeAccumalative, Requests, PrivateMemorySize64, PrivilegedProcessorTimeMilliseconds) VALUES(";

        // StringBuilder pool for better memory reuse
        private StringBuilder _stringBuilder;

        // Disposal flag
        private bool _disposed = false;

        private System.Web.HttpContext moCtx;

        // Session / Request Level Properties
        public System.Web.HttpRequest moRequest;
        public System.Web.HttpResponse moResponse;
        public System.Web.SessionState.HttpSessionState moSession;
        public System.Web.HttpServerUtility moServer;

        public System.Collections.Specialized.NameValueCollection moConfig = (System.Collections.Specialized.NameValueCollection)WebConfigurationManager.GetWebApplicationSection("protean/web");




        public bool Logging
        {
            get
            {
                return bLoggingOn;
            }
            set
            {
                bLoggingOn = value;
            }
        }

        public PerfLog(string SiteName, System.Web.HttpContext oCtx)
        {
            try
            {
                cSiteName = SiteName;
                moCtx = oCtx;

                if (moSession != null)
                {
                    if (moSession["Logging"]?.ToString() == "On")
                    {
                        TurnOn();
                    }
                }
            }
            catch (Exception)
            {
            }
        }

        private void TurnOn()
        {
            // If bLoggingOn Then Exit Sub
            try
            {
                if (!bLoggingOn)
                {
                    if (moCtx != null)
                    {
                        moRequest = moCtx.Request;
                        moResponse = moCtx.Response;
                        if (moCtx.Session != null)
                            moSession = moCtx.Session;
                        moServer = moCtx.Server;
                    }

                    Entries = new List<string>(64); // Start with reasonable capacity
                    _stringBuilder = new StringBuilder(512); // Reusable StringBuilder
                    bLoggingOn = true;
                    nStep = 0;
                    moSession["Logging"] = "On";

                    string cSessionRequest = Convert.ToString(moSession["SessionRequest"]);
                    if (Tools.Number.IsNumeric(cSessionRequest))
                    {
                        moSession["SessionRequest"] = Convert.ToInt16(cSessionRequest) + 1;
                    }
                    else
                    {
                        moSession["SessionRequest"] = 0;
                    }
                    
                    dLast = DateTime.Now;
                    nTimeAccumalative = 0d;
                    nMemLast = 0;
                    nProcLast = 0;

                    // Use shared static counters instead of creating new instances
                    _workingSetPrivateMemoryCounter = _sharedWorkingSetPrivateMemoryCounter.Value;
                    _workingSetMemoryCounter = _sharedWorkingSetMemoryCounter.Value;

                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex.ToString());
            }
        }

        public void Start()
        {
            TurnOn();
        }

        public void Stop()
        {
            bLoggingOn = false;
            moSession["Logging"] = "Off";
            Dispose();
        }

        public void Log(string cModuleName, string cProcessName, string cDescription = "")
        {
            if (!bLoggingOn)
            {
                // Still update LatestLog for memory dumps even when not logging
                LatestLog = $"{cModuleName}-{cProcessName}-{cDescription}";
                return;
            }

            try
            {
                var oLN = DateTime.Now - dLast;
                nTimeAccumalative += oLN.TotalMilliseconds;

                long memoryPrivate = _workingSetPrivateMemoryCounter != null 
                    ? (long)Math.Round(_workingSetPrivateMemoryCounter.NextValue()) 
                    : 0L;

                nMemLast = (int)memoryPrivate;
                nProcLast = Process.GetCurrentProcess().PrivilegedProcessorTime.Milliseconds;

                // Clear and reuse StringBuilder
                _stringBuilder.Clear();
                _stringBuilder.Append(InsertPrefix)
                    .Append('\'').Append(moServer?.MachineName ?? "").Append("','");
                _stringBuilder.Append(cSiteName).Append("','");

                if (moSession?.SessionID != null)
                {
                    _stringBuilder.Append(moSession.SessionID).Append("','");
                    _stringBuilder.Append(moSession["SessionRequest"]?.ToString() ?? "").Append("','");
                }
                else
                {
                    _stringBuilder.Append("','").Append("','");
                }

                string cPath = moCtx?.Request?["Path"] ?? "";

                _stringBuilder.Append(SqlFmt(cPath)).Append("','");
                _stringBuilder.Append(SqlFmt(cModuleName)).Append("','");
                _stringBuilder.Append(TruncateSqlFmt(cProcessName, 254)).Append("','");
                _stringBuilder.Append(TruncateSqlFmt(cDescription, 3999)).Append("',");
                _stringBuilder.Append(nStep).Append(',');
                _stringBuilder.Append(oLN.TotalMilliseconds).Append(',');
                _stringBuilder.Append(nTimeAccumalative).Append(',');
                _stringBuilder.Append(oPerfMonRequests?.RawValue.ToString() ?? "null").Append(",'");
                _stringBuilder.Append(nMemLast).Append("','");
                
                if (_workingSetMemoryCounter != null)
                {
                    _stringBuilder.Append(((long)Math.Round(_workingSetMemoryCounter.NextValue())).ToString());
                }
                
                _stringBuilder.Append("')");

                Entries.Add(_stringBuilder.ToString());
                nStep++;
                dLast = DateTime.Now;
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"PerfLog.Log Error: {ex}");
            }
        }

        /// <summary>
        /// Helper method to truncate SqlFmt results efficiently
        /// </summary>
        private static string TruncateSqlFmt(string value, int maxLength)
        {
            string formatted = SqlFmt(value).ToString();
            return formatted.Length <= maxLength ? formatted : formatted.Substring(0, maxLength);
        }

        public void Write()
        {
            if (!bLoggingOn || Entries == null || Entries.Count == 0) return;
            
            string cProcessInfo = null;
            
            try
            {
                string ConStr = BuildConnectionString();
                
                using (var oCon = new System.Data.SqlClient.SqlConnection(ConStr))
                using (var oCmd = oCon.CreateCommand())
                {
                    oCon.Open();
                    
                    // Batch inserts in groups to avoid SQL Server command limits
                    const int batchSize = 100;
                    var validEntries = Entries.Where(e => !string.IsNullOrEmpty(e)).ToList();
                    
                    for (int i = 0; i < validEntries.Count; i += batchSize)
                    {
                        var batch = validEntries.Skip(i).Take(batchSize);
                        var batchSql = string.Join(";", batch);
                        
                        cProcessInfo = $"Batch {i / batchSize + 1}";
                        oCmd.CommandText = batchSql;
                        oCmd.CommandTimeout = 30; // Explicit timeout
                        
                        try
                        {
                            oCmd.ExecuteNonQuery();
                        }
                        catch (Exception batchEx)
                        {
                            Debug.WriteLine($"PerfLog batch insert failed: {cProcessInfo} - {batchEx}");
                            // Continue with next batch even if this one fails
                        }
                    }
                }
                
                bLoggingOn = false;
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"PerfLog.Write Error: {cProcessInfo} - {ex}");
            }
            finally
            {
                // Clear entries to free memory
                Entries?.Clear();
                Entries = null;
                _stringBuilder = null;
            }
        }

        /// <summary>
        /// Builds the database connection string from config
        /// </summary>
        private string BuildConnectionString()
        {
            string ConStr = moConfig?["PerfMonConnection"];
            if (string.IsNullOrEmpty(ConStr))
            {
                string dbAuth = !string.IsNullOrEmpty(moConfig?["DatabasePassword"])
                    ? $"user id={moConfig["DatabaseUsername"]}; password={moConfig["DatabasePassword"]}"
                    : moConfig?["DatabaseAuth"] ?? "Integrated Security=SSPI;";
                
                ConStr = $"Data Source={moConfig["DatabaseServer"]}; Initial Catalog={moConfig["DatabaseName"]}; {dbAuth}";
            }
            return ConStr;
        }

        #region IDisposable Support
        
        protected virtual void Dispose(bool disposing)
        {
            if (!_disposed)
            {
                if (disposing)
                {
                    try
                    {
                        // Dispose non-shared Performance Counters only
                        if (oPerfMonRequests != null)
                        {
                            oPerfMonRequests.Dispose();
                            oPerfMonRequests = null;
                        }
                        
                        // DO NOT dispose shared static counters - they are reused across requests
                        // Just null out the references
                        _workingSetPrivateMemoryCounter = null;
                        _workingSetMemoryCounter = null;
                        
                        // Clear list (faster than array clearing)
                        if (Entries != null)
                        {
                            Entries.Clear();
                            Entries = null;
                        }
                        
                        // Clear StringBuilder
                        if (_stringBuilder != null)
                        {
                            _stringBuilder.Clear();
                            _stringBuilder = null;
                        }
                        
                        // Clear string to help GC
                        LatestLog = null;
                    }
                    catch (Exception ex)
                    {
                        Debug.WriteLine($"Error disposing PerfLog: {ex}");
                    }
                }
                
                _disposed = true;
            }
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        
        #endregion

    }
}